### DISCUSSION TITLE
"make_service_fn" - gone. Where is it?

### DISCUSSION THREAD

**Discussion by ally9335** [hyperium/hyper#3471] [Q&A] [answered]
**Version**
1.0

**Platform**
any

**Description**
there used to be the function `hyper::service::make_service_fn`

Now it's gone. There's no mention of it in the docs, no in the ChangeLog, nowhere. Where is it?

How to convert this into `v1.0x`?

```
    let sf = make_service_fn(|_connection| async {
        Ok::<_, Infallible>(service_fn(move |req| handle_post_req(req)))
    });
```

**Comment by seanmonstar** [comment, marked as answer]
The `Server` and `make_service_fn` are gone, you should use a `loop { let conn = listener.accept().await; ... }` instead. See https://hyper.rs/guides/1/upgrading/

**Reply by wiseaidev** [reply]
I think it would be best if you could share a simple implementation of this struct, perhaps make it built-in to Hyper so we can import it easily:
https://github.com/hyperium/hyper/blob/8241f91f5bf9bc8def35cc214a5219ad1a39f71e/examples/service_struct_impl.rs#L15
The reason for this is that serving new connections now requires a custom struct that implements `Read, Write, Unpin` traits along with the service:
https://github.com/hyperium/hyper/blob/8241f91f5bf9bc8def35cc214a5219ad1a39f71e/examples/service_struct_impl.rs#L32
https://github.com/hyperium/hyper/blob/8241f91f5bf9bc8def35cc214a5219ad1a39f71e/examples/service_struct_impl.rs#L35

**Comment by EleisonC** [comment]
Hey @ally9335 I found this in https://hyper.rs/guides/0.14/server

**Reply by dswij** [reply]
That guide is for v0.14, while the question is for v1.0. v1.0 introduces breaking changes, one of which is removing `make_service_fn`.

As the answer above stated, this guide is probably answers the question better: https://hyper.rs/guides/1/upgrading/

**Comment by meawoppl** [comment]
FWIW, the upgrading link above does not really discuss this. Perhaps it changed since linked?

**Reply by Kyu** [reply]
`make_service_fn` isn't mentioned in the changelog at all either. 

edit: I guess these are some useful examples: https://hyper.rs/guides/1/server/middleware/
